### PR TITLE
[SPARK-15308][SQL] RowEncoder should preserve nested column name.

### PR DESCRIPTION
**Pull request by ueshin**
## What changes were proposed in this pull request?

The following code generates wrong schema:

```
val schema = new StructType().add(
  "struct",
  new StructType()
    .add("i", IntegerType, nullable = false)
    .add(
      "s",
      new StructType().add("int", IntegerType, nullable = false),
      nullable = false),
  nullable = false)
val ds = sqlContext.range(10).map(l => Row(l, Row(l)))(RowEncoder(schema))
ds.printSchema()
```

This should print as follows:

```
root
 |-- struct: struct (nullable = false)
 |    |-- i: integer (nullable = false)
 |    |-- s: struct (nullable = false)
 |    |    |-- int: integer (nullable = false)
```

but the result is:

```
root
 |-- struct: struct (nullable = false)
 |    |-- col1: integer (nullable = false)
 |    |-- col2: struct (nullable = false)
 |    |    |-- col1: integer (nullable = false)
```

This PR fixes `RowEncoder` to preserve nested column name.
## How was this patch tested?

Existing tests and I added a test to check if `RowEncoder` preserves nested column name.

### CODE REVIEW

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala [hunk 1]
```diff
@@ -149,12 +149,12 @@ object RowEncoder {
         dataType = t)
 
     case StructType(fields) =>
-      val convertedFields = fields.zipWithIndex.map { case (f, i) =>
+      val convertedFields = fields.zipWithIndex.flatMap { case (f, i) =>
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/RowEncoder.scala]: We can follow the style in https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala#L527:

```
val nonNullOutput = CreateNamedStruct(fields.zipWithIndex.flatMap { case (field, index) =>
...
}
```

**Comment by cloud-fan** [comment]
LGTM except one style comment, thanks for working on it!

**Comment by rxin** [comment]
@ueshin unfortunately this no longer merges cleanly ...

**Comment by cloud-fan** [comment]
LGTM, pending jenkins

**Comment by ueshin** [comment]
It seems PySpark tests are broken?

**Comment by ueshin** [comment]
Jenkins, retest this please.

**Comment by rxin** [comment]
Thanks - merging in master/2.0.

**Pull request closed without merging**